examples/stm32/f4/stm32f4-discovery/timer/timer.c

```diff
@@ -0,0 +1,190 @@
+	nvic_enable_irq(NVIC_TIM2_IRQ);
+
+	/* Reset TIM2 peripheral. */
+	timer_reset(TIM2);
+
+	/* Timer global mode:
+	 * - No divider
+	 * - Alignment edge
+	 * - Direction up
+	 */
+	timer_set_mode(TIM2, TIM_CR1_CKD_CK_INT,
+		       TIM_CR1_CMS_EDGE, TIM_CR1_DIR_UP);
+
+	/* Reset prescaler value.
+	 * Running the clock at 5kHz.
+	 */
+	/*
+	 * On STM32F4 the timers are not running directly from pure APB1 or
+	 * APB2 clock busses.  The APB1 and APB2 clocks used for timers might
+	 * be the double of the APB1 and APB2 clocks.  This depends on the
+	 * setting in DCKCFGR register. By default the behaviour is the
+	 * following: If the Prescaler APBx is greater than 1 the derived timer
+	 * APBx clocks will be double of the original APBx frequencies. Only if
+	 * the APBx prescaler is set to 1 the derived timer APBx will equal the
+	 * original APBx frequencies.
+	 *
+	 * In our case here the APB1 is devided by 4 system frequency and APB2
+	 * divided by 2. This means APB1 timer will be 2 x APB1 and APB2 will
+	 * be 2 x APB2. So when we try to calculate the prescaler value we have
+	 * to use rcc_apb1_freqency * 2!!! 
+	 *
+	 * For additional information see reference manual for the stm32f4
+	 * familiy of chips. Page 204 and 213
+	 */
+	timer_set_prescaler(TIM2, ((rcc_apb1_frequency * 2) / 10000));
+
+	/* Enable preload. */
+	timer_disable_preload(TIM2);
+
+	/* Continous mode. */
+	timer_continuous_mode(TIM2);
+
+	/* Period (36kHz). */
+	timer_set_period(TIM2, 65535);
+
+	/* Disable outputs. */
+	timer_disable_oc_output(TIM2, TIM_OC1);
+	timer_disable_oc_output(TIM2, TIM_OC2);
+	timer_disable_oc_output(TIM2, TIM_OC3);
+	timer_disable_oc_output(TIM2, TIM_OC4);
+
+	/* -- OC1 configuration -- */
+
+	/* Configure global mode of line 1. */
+	timer_disable_oc_clear(TIM2, TIM_OC1);
+	timer_disable_oc_preload(TIM2, TIM_OC1);
+	timer_set_oc_slow_mode(TIM2, TIM_OC1);
+	timer_set_oc_mode(TIM2, TIM_OC1, TIM_OCM_FROZEN);
+
+	/* Set the capture compare value for OC1. */
+	timer_set_oc_value(TIM2, TIM_OC1, 1000);
+
+	/* ---- */
+
+	/* ARR reload enable. */
+	timer_disable_preload(TIM2);
+
+	/* Counter enable. */
+	timer_enable_counter(TIM2);
+
+	/* Enable commutation interrupt. */
+	timer_enable_irq(TIM2, TIM_DIER_CC1IE);
+}
+
+void tim2_isr(void)
+{
+	if (timer_get_flag(TIM2, TIM_SR_CC1IF)) {
+
+		/* Clear compare interrupt flag. */
+		timer_clear_flag(TIM2, TIM_SR_CC1IF);
+
+		/*
+		 * Get current timer value to calculate next
+		 * compare register value.
+		 */
+		compare_time = timer_get_counter(TIM2);
+
+		/* Calculate and set the next compare value. */
+		frequency = frequency_sequence[frequency_sel++];
+		new_time = compare_time + frequency;
+
+		timer_set_oc_value(TIM2, TIM_OC1, new_time);
+		if (frequency_sel == 18)
+			frequency_sel = 0;
+
+		/* Toggle LED to indicate compare event. */
+		gpio_toggle(GPIOD, GPIO12);
+		gpio_toggle(GPIOD, GPIO13);
+	}
+}
+
+int main(void)
+{
+	clock_setup();
+	gpio_setup();
+	tim_setup();
+
+	while (1)
+		__asm("nop");
```

Comment:
You mean __WFI();? I did not know about this assembly call. Good to know. Thanks for pointing it out. I will use that. For those reading this. To save you time searching what this is. It stands for "Wait For Interrupt" and is a hint instruction that suspends execution until one of the following events occurs:

* a non-masked interrupt occurs and is taken
* an interrupt masked by PRIMASK becomes pending
* a Debug Entry request

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0552a/BABFEFIG.html